src/Gui/ViewProviderCoordinateSystem.cpp

```diff
@@ -83,6 +85,50 @@ void ViewProviderCoordinateSystem::attach(App::DocumentObject* pcObject)
     addDisplayMaskMode(pcGroupChildren, "Base");
 }
 
+void ViewProviderCoordinateSystem::finishRestoring()
+{
+    showMigrationDialog();
+}
+
+void ViewProviderCoordinateSystem::showMigrationDialog()
+{
+    auto lcs = dynamic_cast<App::LocalCoordinateSystem*>(getObject());
+    if (!lcs || !lcs->migrated) {
+        return;
+    }
+
+    static bool userWarned = false;
+
+    if (userWarned || !App::GetApplication().GetParameterGroupByPath(
+        "User parameter:BaseApp/Preferences/View")->GetBool("ShowLCSMigrationWarning", true)) {
+        return;
+    }
+
+    // Display the warning message
+    QMessageBox msgBox(QMessageBox::Warning,
+        QObject::tr("File Migration Warning"),
+        QObject::tr("This file was created with an older version of FreeCAD. "
+            "Origin axes had incorrect placements, which have now been corrected.\n\n"
```

Comment:
IMO the first part of this message can lead to confusion because it sounds like a bug in older FreeCAD versions. But in fact there was no bug but a different behaviour between the various coordinate system implementations.

Reply:
Yes it is right but I wasn't sure how to word it.

Reply:
Maybe something like "Historically origin axis had inconsistent placement which is being fixed now ...

Reply:
Another problem of this whole function is that the dialog is only shown once for a whole FreeCAD session. So, this means once the dialog was shown and the user opens another file that must be migrated the user won't be informed about this any more.

A better way would be to set a status flag to the document (similar to Document::PartialRestore or Document::RestoreError) and if needed add a std::string to hold a custom message. The dialog handling then can be done in StdCmdOpen::activated or Application::open.

Another positive side-effect is that it will stop blocking the application during the integration tests.